connector/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryDataWriterHelper.java

```diff
@@ -0,0 +1,364 @@
+import com.google.cloud.bigquery.storage.v1.stub.readrows.ApiResultRetryAlgorithm;
+import com.google.cloud.bigquery.storage.v1alpha2.*;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Int64Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.*;
+
+/**
+ * The interface which sketches out the necessary functions in order for a Spark DataWriter to
+ * interact with the BigQuery Storage Write API.
+ */
+interface BigQueryDataWriterHelper {
+
+  static BigQueryDataWriterHelper from(
+      BigQueryWriteClientFactory writeClientFactory,
+      String tablePath,
+      ProtoBufProto.ProtoSchema protoSchema,
+      RetrySettings bigqueryDataWriterHelperRetrySettings) {
+    return new BigQueryDataWriterHelperDefault(
+        writeClientFactory, tablePath, protoSchema, bigqueryDataWriterHelperRetrySettings);
+  }
+
+  void addRow(ByteString message) throws IOException;
+
+  long commit() throws IOException;
+
+  String getWriteStreamName();
+
+  void abort();
+}
+
+/** The current default implementation of such a helper interface. */
+class BigQueryDataWriterHelperDefault implements BigQueryDataWriterHelper {
+
+  final Logger logger = LoggerFactory.getLogger(BigQueryDataWriterHelperDefault.class);
+
+  // 1MB soft limit for each append:
+  final long SOFT_LIMIT_APPEND_ROWS_REQUEST_SIZE = 1000L * 1000L;
+  // 2MB hard limit for an append (in the case  a single row exceeds
+  // SOFT_LIMIT_APPEND_ROWS_REQUEST_SIZE):
+  final long HARD_LIMIT_APPEND_ROWS_REQUEST_SIZE = 2 * 1000L * 1000L;
+
+  private final BigQueryWriteClient writeClient;
+  private final String tablePath;
+  private final ProtoBufProto.ProtoSchema protoSchema;
+  private final RetrySettings retrySettings;
+
+  private String writeStreamName;
+  private StreamWriter streamWriter;
+  private ProtoBufProto.ProtoRows.Builder protoRows;
+
+  private long appendRequestRowCount = 0; // number of rows waiting for the next append request
+  private long appendRequestSizeBytes = 0; // number of bytes waiting for the next append request
+
+  private long writeStreamSizeBytes = 0; // total bytes of the current write-stream
+  private long writeStreamRowCount = 0; // total offset / rows of the current write-stream
+
+  BigQueryDataWriterHelperDefault(
+      BigQueryWriteClientFactory writeClientFactory,
+      String tablePath,
+      ProtoBufProto.ProtoSchema protoSchema,
+      RetrySettings bigqueryDataWriterHelperRetrySettings) {
+    this.writeClient = writeClientFactory.createBigQueryWriteClient();
+    this.tablePath = tablePath;
+    this.protoSchema = protoSchema;
+    this.retrySettings = bigqueryDataWriterHelperRetrySettings;
+
+    try {
+      this.writeStreamName = retryCreateWriteStream();
+    } catch (ExecutionException | InterruptedException e) {
+      throw new RuntimeException("Could not create write-stream after multiple retries.", e);
```

Comment:
Please change `RuntimeException` to a subclass with better description
nit: remove dot in the end of the message